### PR TITLE
Discusses #7

### PR DESCRIPTION
**Pull request by klngrs**
Looking for assistance getting the rpc param to actually rerender the widget. Not sure of the redirectMap schema...

Context: https://github.com/NEARBuilders/near-bos-webcomponent/issues/7
Demo: https://cloudflare-ipfs.com/ipfs/QmS1AryveoGbC4BWD4wm1CcxkbzrGJDfGS5BVS9HbK1i8P

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -39,6 +39,7 @@ function Viewer({ widgetSrc, code, initialProps }) {
   useEffect(() => {
```
Comment (elliotBraem) [inline src/App.js:39]: RPC solution is to replace the need for this useEffect

Currently, this useEffect fetches the redirectMap served from bos-workspace via a url in local storage

File: src/App.js [hunk 2]
```diff
@@ -99,7 +103,8 @@ function App(props) {
         },
         config: {
           defaultFinality: undefined,
-        },
+          ...(props.rpc ? { nodeUrl: props.rpc } : {})
+        }
```
Comment (elliotBraem) [inline src/App.js:107]: This is correct, although let's make it simpler with a default RPC. I like lava network as default.
We should also make this dependent on networkId (testnet or mainnet).

nodeUrl: props.rpc || (networkId === "mainnet" ? "https://near.lava.build" : "https://near-testnet.lava.build")

File: src/index.js [hunk 3]
```diff
@@ -22,19 +22,20 @@ class NearSocialViewerElement extends HTMLElement {
...
     static get observedAttributes() {
-        return ['src', 'code', 'initialProps'];
+        return ['src', 'code', 'initialProps', 'rpc'];
     }
 
     attributeChangedCallback(name, oldValue, newValue) {
         // Re-render the component with new props when attributes change
         if (oldValue !== newValue) {
-            const container = this.shadowRoot.querySelector('div');
+            const container = this.shadowRoot.querySelector('slot');
```
Comment (elliotBraem) [inline src/index.js:38]: What does this do?

File: src/App.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ import {
 
 const SESSION_STORAGE_REDIRECT_MAP_KEY = "nearSocialVMredirectMap";
 
-function Viewer({ widgetSrc, code, initialProps }) {
+function Viewer({ widgetSrc, code, initialProps, rpc }) {
   const location = useLocation();
   const searchParams = new URLSearchParams(location.search);
   const [redirectMap, setRedirectMap] = useState({});
```
Comment (elliotBraem) [inline src/App.js:22]: I think we can remove the redirectMap entirely from here

**Review by elliotBraem — COMMENTED** [review]
Left comments -- I think this change in the near-bos-webcomponent looks good after removing the useEffect for redirectMap.

When we try to load widgets via something like `<Widget src="efiz.near/widget/Tree`, we are fetching the code from on chain via an RPC request in the VM.  Currently, we have this check for "redirectMap" in the VM, which checks if the path (efiz.near/widget/Tree) exists in our redirectMap so we use that, otherwise it makes the RPC request.

My thought has been that, during local development, we would pass in the local bos-workspace rpc address, which serves as a proxy to true RPC. So rather than the VM checking redirectMap, our bos-workspace server intercepts the request and checks redirectMap it has. This change would probably happen here: https://github.com/NEARBuilders/bos-workspace/blob/6df6babdaa88239700298a0054cefaf83245d553/lib/dev.ts#L160 

@petersalomonsen would you agree with this?

**Comment by petersalomonsen** [comment]
> Left comments -- I think this change looks good after removing the useEffect for redirectMap.
> 
> For context -- when we try to load widgets in code via `<Widget src="efiz.near/widget/Tree`, we are fetching the code from on chain via [an RPC request in the VM](https://github.com/NearSocial/VM/blob/8dbb06526f15f10d562a118a38e4c0b6412beb88/src/lib/components/Widget.js#L93). Although during local development, rather than send an rpc request, we first check if this src ("efiz.near/widget/Tree") key exists in the "redirectMap" served from bos-workspace. If the key exists, we get the code from the map -- this check happens in the VM [here](https://github.com/NearSocial/VM/blob/8dbb06526f15f10d562a118a38e4c0b6412beb88/src/lib/data/utils.js#L420).
> 
> I think the next step in this solution would be a change to bos-workspace. So that during local development, we would pass a local rpc address instead, coming from the local bos-workspace server, to act as a proxy to true RPC. The check against the redirectMap, which bos-workspace is building, will happen here, rather than in the VM. This change would probably happen here: [bos-workspace/lib/dev.ts](https://github.com/NEARBuilders/bos-workspace/blob/6df6babdaa88239700298a0054cefaf83245d553/lib/dev.ts#L160)
> 
> @petersalomonsen would you agree with this?

The same approach is actually used for playwright tests in DevHub as you can see here: https://github.com/NEAR-DevHub/neardevhub-bos/blob/main/playwright-tests/util/bos-loader.js. So yes, going through an RPC proxy should be a good approach.

Let us figure out the appropriate way to configure this for near-bos-webcomponent.

File: src/index.js (hunk 3 above)
Comment (klngrs) [reply src/index.js:38]: For some reason, it wasn't finding the div in the shadow root when it went to re-render. But querying for the slot picked up the node it was looking to render. It appeared to have the same effect. Not married to it, but it did enable the re-render to work (showing the rpc param as the attribute)

**Comment by klngrs** [comment]
Cool, thanks all. And thanks @petersalomonsen for the redirectMap explanation. The location of where the redirect was happening was escaping me, but this makes more sense now. 

Looks like @elliotBraem is applying the proxy as middleware in the server gateway code.

**Comment by elliotBraem** [comment]
bos-workspace now supports rpc url via https://github.com/NEARBuilders/bos-workspace/pull/80

`URL: http://127.0.0.1:8080/api/proxy-rpc`

**Comment by petersalomonsen** [comment]
would be great with a playwright test that instantiate the webcomponent with and without the rpc option, and with different RPCs. That would be a live example of how to use it.

**Comment by elliotBraem** [comment]
I've opened up a PR that builds off of this one, (please review #11) -- @klngrs I've invited you as a collaborator 

I included two tests -- one without the rpc option and one with a custom value. 

It works when I input the rpc url manually within the app itself, but not when it is passed via the parameter. It seems that rpc is initially undefined during connectedCallback, so it falls back to default, then it sets correctly with attributeChangedCallback (see logs), but the default rpc is already being used.

Because of this custom rpc url test fails, but default test passes -- any ideas on to fix this?

**Comment by elliotBraem** [comment]
Closing this discussion with merging #11

**Pull request closed without merging**